tfhe/src/c_api/high_level_api/threading.rs

```diff
@@ -0,0 +1,114 @@
+use super::super::utils::{catch_panic, get_mut_checked};
+use crate::c_api::high_level_api::keys::ServerKey;
+use crate::c_api::utils::get_ref_checked;
+use rayon::{ThreadPool, ThreadPoolBuilder};
+use std::ffi::{c_int, c_void};
+
+/// Used to limit number of threads
+///
+/// This struct can be used to limit the number of threads
+/// operations (ran inside this context) can use.
+///
+/// A threading context creates and hold the specified number of threads
+pub struct TfheThreadingContext {
+    pool: ThreadPool,
+}
+
+/// Creates a new threading context
+///
+/// - num_threads: number of threads inside this context 0 means it will number of CPU threads
+#[no_mangle]
+pub unsafe extern "C" fn tfhe_threading_context_create(
+    num_threads: usize,
+    context: *mut *mut TfheThreadingContext,
+) -> c_int {
+    catch_panic(|| {
+        *context = std::ptr::null_mut();
+
+        let pool = ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .build()
+            .unwrap();
+
+        let ctx = TfheThreadingContext { pool };
+
+        *context = Box::into_raw(Box::new(ctx));
+    })
+}
+
+/// Destroys the threading context
+///
+/// `context` may be NULL
+#[no_mangle]
+pub unsafe extern "C" fn tfhe_threading_context_destroy(context: *mut TfheThreadingContext) {
+    let _ = catch_panic(|| {
+        if context.is_null() {
+            return;
+        }
+
+        let _ = Box::from_raw(context);
```

Comment:
I think I had a clippy warning (a while ago) saying it was better to do drop(Box::from_raw(...)) to be more explicit 🤔